pkg/util/vsphere/vm/cdrom.go

```diff
@@ -0,0 +1,217 @@
+			} else if sata := curDevices.PickController((*vimtypes.VirtualSATAController)(nil)); sata != nil {
+				curDevices.AssignController(cdrom, sata)
+			} else {
+				ahci, controllerChanges := createNewAHCIController(curDevices)
+				deviceChanges = append(deviceChanges, controllerChanges...)
+				curDevices.AssignController(cdrom, ahci)
+			}
+
+			// Add the new CD-ROM device to the current devices list to ensure
+			// new devices are added to the VM with proper keys.
+			curDevices = append(curDevices, cdrom)
+			deviceChanges = append(deviceChanges, &vimtypes.VirtualDeviceConfigSpec{
+				Device:    cdrom,
+				Operation: vimtypes.VirtualDeviceConfigSpecOperationAdd,
+			})
+		} else {
+			// Existing CD-ROM device found with the current backing file.
+			// Update the device's connection state if changed.
+			deviceChanges = append(deviceChanges, updateCdromConnect(curCdrom, !specCdrom.Disconnect)...)
+		}
+	}
+
+	// Delete existing CD-ROMs whose backing file not in the expected list.
+	for _, d := range curDevices {
+		if b, ok := d.GetVirtualDevice().Backing.(*vimtypes.VirtualCdromIsoBackingInfo); ok {
+			if _, ok := expectedBackingFileNames[b.FileName]; !ok {
+				vmCtx.Logger.Info("Deleting unexpected CD-ROM device", "cdrom", d)
+				deviceChanges = append(deviceChanges, &vimtypes.VirtualDeviceConfigSpec{
+					Device:    d.GetVirtualDevice(),
+					Operation: vimtypes.VirtualDeviceConfigSpecOperationRemove,
+				})
+			}
+		}
+	}
+
+	return deviceChanges, nil
+}
+
+func getIsoImageFileName(
+	vmCtx pkgctx.VirtualMachineContext,
+	client ctrlclient.Client,
+	imageName, imageKind string) (string, error) {
+	var itemStatus imgregv1a1.ContentLibraryItemStatus
+	switch imageKind {
+	case vmiKind:
+		ns := vmCtx.VM.Namespace
+		var vmi vmopv1.VirtualMachineImage
+		if err := client.Get(vmCtx, ctrlclient.ObjectKey{Name: imageName, Namespace: ns}, &vmi); err != nil {
+			return "", err
+		}
+		var clitem imgregv1a1.ContentLibraryItem
+		if err := client.Get(vmCtx, ctrlclient.ObjectKey{Name: vmi.Spec.ProviderRef.Name, Namespace: ns}, &clitem); err != nil {
+			return "", err
+		}
+		itemStatus = clitem.Status
+	case cvmiKind:
+		var cvmi vmopv1.ClusterVirtualMachineImage
+		if err := client.Get(vmCtx, ctrlclient.ObjectKey{Name: imageName}, &cvmi); err != nil {
+			return "", err
+		}
+		var cclitem imgregv1a1.ClusterContentLibraryItem
+		if err := client.Get(vmCtx, ctrlclient.ObjectKey{Name: cvmi.Spec.ProviderRef.Name}, &cclitem); err != nil {
+			return "", err
+		}
+		itemStatus = cclitem.Status
+	}
+
+	if itemStatus.Type != imgregv1a1.ContentLibraryItemTypeIso {
+		return "", fmt.Errorf("expected ISO type image, got %s", itemStatus.Type)
+	}
+	if len(itemStatus.FileInfo) == 0 || itemStatus.FileInfo[0].StorageURI == "" {
+		return "", fmt.Errorf("no storage URI found in the content library item of image")
+	}
+
+	uri := itemStatus.FileInfo[0].StorageURI
+	if strings.Contains(uri, serverIDSuffix) {
```

Comment:
With StorageURI currently being:
```
${datastore.info.url}/contentlib-${lib_id}/${item_id}/ubuntu-impish-21.10-cloudimg_${file_id}.ovf?serverId=${server_id}
```
We need to strip out `?serverId=${server_id}` here. And vCenter converts the datastore info.url to name:
```
[sharedVmfs-0] contentlib-${lib_id}/${item_id}/ubuntu-impish-21.10-cloudimg_${file_id}.ovf
```
Should we just change image-registry-operator's FileInfo[].StorageURI to to same format as above?
It would be easier to read and usable for file backing as-is. I can open an MR for the change.

Note, can find datastore.info.url with this property:
```console
% govc object.collect -s /*.wcp-sanity/datastore/sharedVmfs-0 info.url
ds:///vmfs/volumes/666d7a79-cb0d28b2-57c8-0645602e1b58/
```

Reply:
Yep, I added a TODO comment to address this once it's supported in govmomi.